Variable Resolution and Caching

sschmittBt opened the pull request:
- added an IFieldValueResolver to query execution that resolves any variable field values of the search condition
- added IVariableFieldCondition interface and implementation that marks a field condition as having a variable value
- changed search condition creation in the query to decide if a field condition is fixed or variable
- removed limit and offset from the query. They are now instead given as parameters during query execution
- added default query limit to the datastore to account for removal of the default limit value inside the query
- added cache to IFieldCondition to cache any intermediate results created during the transformation to the IQueryExpression
- check the IFieldCondition cache during the building of the IQueryExpression to speed up the build

sschmittBt committed:
Variable Resolution and Caching
-added an IFieldValueResolver to query execution that resolves any variable field values of the search condition
-added IVariableFieldCondition interface and implementation that marks a field condition as having a variable value
-changed search condition creation in the query to decide if a field condition is fixed or variable
-removed limit and offset from the query. They are now instead given as parameters during query execution
-added default query limit to the datastore to account for removal of the default limit value inside the query
-added cache to IFieldCondition to cache any intermediate results created during the transformation to the IQueryExpression
-check the IFieldCondition cache during the building of the IQueryExpression to speed up the build

sschmittBt committed:
Fixed method call

sschmittBt committed:
Merged 'develop'.

Commit:
Merge branch 'pojomapper_pom_parent' into feature/#53-Query-VariableResolution-and-Caching

remi128 commented on vertx-pojo-mapper-common-test/src/main/java/de/braintags/io/vertx/pojomapper/dataaccess/query/TestFieldConditionCache.java:
```diff
@@ -0,0 +1,115 @@
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+
+public class TestFieldConditionCache extends DatastoreBaseTest {
```
Tests to be executes must be contained inside a TestSuite, here probably in AllTestCommon

sschmittBt replied:
Added to AllTestCommon

remi128 commented on vertx-pojo-mapper-common-test/src/main/java/de/braintags/io/vertx/pojomapper/testdatastore/DatastoreBaseTest.java:
```diff
@@ -44,10 +44,10 @@
  * on how to init
- * 
- * 
+ *
```
seems that we are still having whitespace difference. WE should clear that out to avoid trashy diffs

sschmittBt replied:
I accidentally had "remove whitespace at end of lines" enabled as save format action. Disabled it.

remi128 commented on vertx-pojo-mapper-common/src/main/java/de/braintags/io/vertx/pojomapper/dataaccess/query/IFieldValueResolver.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Resolves any variable inside a {@link IFieldCondition} with an actual object. The value can come from anywhere,
+ * routing context, request parameter, field content of another POJO...
+ *
```
please add a concrete example into documentation

sschmittBt replied:
Added example

remi128 commented on vertx-pojo-mapper-common/src/main/java/de/braintags/io/vertx/pojomapper/dataaccess/query/IQuery.java:
```diff
@@ -31,46 +31,64 @@
+   * @param resultHandler
+   *          contains the {@link IQueryResult}
    */
-  IQuery<T> setLimit(int limit);
```
where is the limit defined now? Imo its missing, or?

sschmittBt replied:
It's given as parameter to the execution directly:
void execute(IFieldValueResolver resolver, int limit, int offset,
      Handler<AsyncResult<IQueryResult<T>>> resultHandler);

Query still has another method without limit that uses the new default value of the datastore.

remi128 commented on vertx-pojo-mapper-common/src/main/java/de/braintags/io/vertx/pojomapper/dataaccess/query/IQuery.java:
```diff
@@ -31,46 +31,64 @@
+   * @param resultHandler
+   *          handler for the completed {@link IQueryExpression}
    */
-  IQuery<T> setStart(int start);
```
missing either?

remi128 commented on vertx-pojo-mapper-common/src/main/java/de/braintags/io/vertx/pojomapper/dataaccess/query/impl/FieldCondition.java:
```diff
@@ -12,6 +12,9 @@
+import java.util.HashMap;
+import java.util.Map;
+
 import org.apache.commons.lang3.StringUtils;
```
sure that StringUtils are not blocking?

sschmittBt replied:
Blocking what? I just use them for the toString if the value is a list. Doing it manually, I would still have to iterate over the list, which is also blocking.

remi128 commented on vertx-pojo-mapper-common/src/main/java/de/braintags/io/vertx/pojomapper/dataaccess/query/impl/Query.java:
```diff
@@ -480,4 +461,26 @@ public ISearchConditionContainer and(ISearchCondition... searchConditions) {
+  private IFieldCondition createFieldCondition(String field, QueryOperator operator, Object value) {
+    if (value instanceof String && StringUtils.isNotBlank((String) value)) {
+      String stringValue = (String) value;
+      if (stringValue.startsWith("${") && stringValue.endsWith("}")) {
```
can occur something like " ${variable} "?

sschmittBt replied:
Well, that would be incorrect syntax. I could trim the value, but I don't know if there should really be much more operations during a string check, since most non-variable values will also be string, so we would trim unnecessarily most of the time.

remi128 commented on vertx-pojo-mapper-common/src/main/java/de/braintags/io/vertx/pojomapper/impl/AbstractDataStore.java:
```diff
@@ -60,11 +61,12 @@ public AbstractDataStore(Vertx vertx, JsonObject properties) {
     this.vertx = vertx;
     this.properties = properties;
     initSupportedKeyGenerators();
+    defaultQueryLimit = properties.getInteger("defaultQueryLimit", 500);
```
replace expression "defaultQueryLimit" by well documented constant in IDataStore

sschmittBt replied:
Added constant.

sschmittBt committed:
Merged 'develop'.

# Conflicts:
#	vertx-pojo-mapper-common-test/src/main/java/de/braintags/vertx/jomnigate/testdatastore/DatastoreBaseTest.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryExpression.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractQueryResult.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/AbstractSearchConditionContainer.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/IQueryExpression.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/Query.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/dataaccess/query/impl/SortDefinition.java
#	vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/impl/AbstractDataStore.java
#	vertx-pojo-mapper-json/src/test/java/de/braintags/vertx/jomnigate/impl/DummyDataStore.java
#	vertx-pojo-mapper-mysql/src/main/java/de/braintags/vertx/jomnigate/mysql/dataaccess/SqlDelete.java
#	vertx-pojo-mapper-mysql/src/main/java/de/braintags/vertx/jomnigate/mysql/dataaccess/SqlQuery.java
#	vertx-pojongo/src/main/java/de/braintags/vertx/jomnigate/mongo/dataaccess/MongoDelete.java
#	vertx-pojongo/src/main/java/de/braintags/vertx/jomnigate/mongo/dataaccess/MongoQuery.java

sschmittBt committed:
Change requests
- added field condition cache test to test suite
- moved default query limit property to constant
- added example to FieldValueResolver and extended javadoc of IVariableFieldCondition
- fixed comment for query execution

sschmittBt committed:
Classpath changes
-changed paths in comments to reflect the new classpaths
